batchflow/sampler.py

```diff
@@ -266,63 +198,174 @@ def truncate(self, high=None, low=None, expr=None, prob=0.5):
         Sampler
             new Sampler-instance, truncated version of self.
         """
-        # clean up truncation params
-        if high is not None:
-            high = np.array(high).reshape(1, -1)
-        if low is not None:
-            low = np.array(low).reshape(1, -1)
-
-        def truncated(size):
-            """ Truncated sampling method.
-            """
-            if size == 0:
-                return self.sample(size=0)
-
-            # set batch-size
-            expectation = size / prob
-            sigma = np.sqrt(size * (1 - prob) / (prob**2))
-            batch_size = int(expectation + 2 * sigma)
-
-            # sample, filter out, concat
-            ctr = 0
-            cumulated = 0
-            samples = []
-            while cumulated < size:
-                # sample points and compute condition-vector
-                sample = self.sample(size=batch_size)
-                cond = np.ones(shape=batch_size).astype(np.bool)
-                if low is not None:
-                    if expr is not None:
-                        cond &= np.greater_equal(expr(sample).reshape(batch_size, -1), low).all(axis=1)
-                    else:
-                        cond &= np.greater_equal(sample, low).all(axis=1)
-
-                if high is not None:
-                    if expr is not None:
-                        cond &= np.less_equal(expr(sample).reshape(batch_size, -1), high).all(axis=1)
-                    else:
-                        cond &= np.less_equal(sample, high).all(axis=1)
-
-                if high is None and low is None:
-                    cond &= expr(sample).all(axis=1)
-
-                # check that truncation-prob is not to small
-                _share = np.sum(cond) / batch_size
-                if _share < SMALL_SHARE and ctr > 0:
-                    raise ValueError('Probability of region of interest is too small. Try other truncation bounds')
-
-                # get points from region of interest
-                samples.append(sample[cond])
-                cumulated += np.sum(cond)
-                ctr += 1
-
-            return np.concatenate(samples)[:size]
-
-        # init new Sampler, define its sampling-method
-        result = Sampler()
-        result.sample = truncated
-
-        return result
+        return TruncateSampler(self, high, low, expr, prob)
+
+
+class OrSampler(Sampler):
+    def __init__(self, left, right, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.left = left
+        self.right = right
+
+        # calculate probs of samplers in mixture
+        _ws = np.array([self.left.weight, self.right.weight])
+        self.weight = np.sum(_ws)
+        self._normed = _ws / np.sum(_ws)
+
+    def sample(self, size):
+        """ Sampling procedure of a mixture of two samplers.
```

Comment:
This is not an informative description: readers already know that this is a class to create mixture of samplers. A documentation of what is actually going on would help